Basic Structure edits + updates

Cogno-Marco opened the pull request:
## Premise
In this pr we're proposing part of the work we're doing for the replicated network.
Since it's only part of the work, most of the stuff is a work in progress, so we're using this pr to let you updated on all the work we're doing, while also discussing the architecture we're going with, so maybe someone can give some suggestions on how to improve this.

Overall, there's not much complicated, so don't be afraid of the 700+ lines of edits, you'll quickly find out that most of that is trivial code, thanks to the proposed structure or due to how java/android-studio work.

## Main Changes

### 1. Command Design Pattern
While thinking about everything that happens in our replicated network, one thing popped up immediatly.
The chain of events of some commands is very similar for 4 commands, these being `addPeer`, `addResource`, `removePeer` and `removeResource`.
These 4 are basically doing the same chain of events but for different functions
(for example `addPeer` calls an `addPeerLocal` and a `broadcastAddPeer`, same for `removePeer` etc.)
Basically these 4 commands can be seen as
```
x -> xLocal
x -> broadcastx
```
When looking at this one design pattern instantly popped to mind, which is the [Command Design Pattern](https://refactoring.guru/design-patterns/command).

As you can see with files like `CommandExecutor`, `addPeerCommand` etc. we decided to adopt this pattern.

This helps us by, first, separating the code in different structures (commands which start the work and handlers which offer some work, in the same way that `addPeerCommand` has code to execute a task, `broadcast` has code to offer some work, namely broadcast a message), and second, by having smaller code it's a lot easier to modify it, test it, and overrall mantain it.

This whole work is being done by the `CommandExecutor` and the 4 Command classes (`AddPeer`, `AddResource`, `RemovePeer` and `RemoveResource`) thanks to the `Command` interface.

### 2. NetworkDictionary
A basic implementation of the network dictionary has been created and tested, the Dictionary is then mantained inside SMSNetworkManager

### 3. NetworkSubscribers
A basic implementation of the network subscribers list has been created and tested, the list is then mantained inside SMSNetworkManager

### 4. SMSNetwork class structure
A basic structure of the SMSNetwork classes (`SMSNetworkManager` and `SMSJoinableNetManager`) have been implemented (with specifications from the interfaces recently agreed upon, see #6 for details), though these classes are still empty, so this is a trivial edit.

## Side changes
Other than those "big" features, there has been a number of trivial side changes, which you can quickly skip, if you want, these being:

### 1. Listeners Folder
Listener Interfaces agreed on #6 have been moved to a proper folder, because of these a lot of classes edits are only missing imports (these classes being `GetResourceListener`, `InviteListener`, `JoinInvitationListener`, `RemoveResourceListener`, `SetResourceListener` and `NetworkManager`)

### 2. Missing Test Folders
Test folders were missing when work has been started, due to this the class `ExampleInstrumentedTest` was re-introduced with it's own folder.

### 3. `misc.xml`
Android Studio automatically generated file was changed, I'd like a suggestion on how to not let this file change again, but for now everything seems to be working as expected

### 4. unit-tests
A bunch of unit-tests have been created along with the features added (like the list of subscribers and the resource dictionary)
___
Sorry again for the tons of lines/files, but again, most of it is trivial. Happy review 🎉

giovannivelludo committed:
Merge pull request #1 from Cogno-IDU/master

Merge "Cogno-IDU/network-dictionary" into "EIS0/network-dictionary"

giovannivelludo committed:
fixed .gitignore

giovannivelludo committed:
Delete modules.xml

giovannivelludo committed:
Merge pull request #2 from EIS0/hotfix/gitignore

Hotfix/gitignore

giovannivelludo committed:
Merge pull request #3 from Cogno-IDU/master

fixed .gitignore (#3)

giovannivelludo committed:
updated submodule library

giovannivelludo committed:
Merge remote-tracking branch 'upstream/master'

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

Cogno-Marco committed:
Added missing tests folders

Cogno-Marco committed:
Defined interface for NetworkDictionary

Cogno-Marco committed:
Improved Specifications
+ Added unit tests in SMSNetDictionaryTest
- Removed example test class
~ Improved specifications for SMSNetDictionary

Cogno-Marco committed:
Implemented all NetworkDictionary functions, every test was passed

Cogno-Marco committed:
Defined interface for NetSubscribers

Cogno-Marco committed:
Unit tests
~ Improved specifications
+ Created unit tests

Cogno-Marco committed:
Implemented SMSNetSubscribers class, every unit test was passed

Cogno-Marco committed:
Improved SMSNetDictionary constructor

Cogno-Marco committed:
Merge branch 'feature/subscribers' into develop

EnricoCestaro committed:
Added Commands

EdoardoRaimondi committed:
Command interface added

EdoardoRaimondi committed:
CommandExecutor class added

EnricoCestaro committed:
Implemented interface

EnricoCestaro committed:
Correcting a mistake (of mine)

EnricoCestaro committed:
Basic Commands structure completed

EdoardoRaimondi committed:
Commit to move to an apposite branch

EdoardoRaimondi committed:
Merge branch 'develop' of https://github.com/EIS0/network-dictionary into develop

# Conflicts:
#	networklibrary/src/main/java/com/eis/communication/network/Commands/AddPeer.java

EdoardoRaimondi committed:
AddPeer command created

EdoardoRaimondi committed:
Subscribers commands added, with test

EdoardoRaimondi committed:
Add and remove resources commands added (with tests)

Cogno-Marco committed:
+ Defined specifications for SMSNetworkManager and SMSJoinableNetManager
+ implemented acceptJoin method structure using inheritance

Cogno-Marco committed:
Added missing todos

Cogno-Marco committed:
~ Moved test in proper test folder
~ Fixed InvalidTelephone exceptions in a couple of tests

Cogno-Marco committed:
Moved listener interfaces to a proper folder

Cogno-Marco committed:
Fixed import errors

CremaLuca commented:
> `misc.xml`  I'd like a suggestion on how to not let this file change again, but for now everything seems to be working as expected

How about placing it inside the `.gitignore`?

CremaLuca commented on .idea/misc.xml:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" project-jdk-name="1.8" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_7" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
Nah let's keep it 1.8

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/SMSJoinableNetManager.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @author Marco Cognolato
+ */
+public class SMSJoinableNetManager extends SMSNetworkManager
```
You're using basic `FailReason` as fail reason list, couldn't be that there are more reasons why a sms network might fail? Did you think about this (adding a `SMSReplicatedNetFailReason` class)? Are you already willing to do this in the future?

Cogno-Marco replied:
I'm definitely gonna change FailReason to a more inclusive one, I just didn't want to add to much stuff to an already pretty big pr.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/INetworkDictionary.java:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @author Marco Cognolato
+ */
+public interface INetworkDictionary<K,R>{
```
Ew, `I` before interfaces name, how C#-ish.

Cogno-Marco replied:
In fact I come from a mainyl C# language. Anyway, I think using the `I` would be better because you can instantly distinguish an interface from a class.
I mean, would you easily understand that `NetworkManager` was an interface?

CremaLuca replied:
https://stackoverflow.com/questions/2814805/java-interfaces-implementation-naming-convention
> Putting I in front is just Hungarian style notation tautology that adds nothing but more stuff to type to your code

https://howtodoinjava.com/java/basics/java-naming-conventions/

I'm not a fan of "I" at all. Also only these two interfaces changed name? Either you change them all (shivers down my back) or you call this propely by removing the I.

Cogno-Marco replied:
> adds nothing but more stuff to type to your code

1 letter, lol
Also no, gives more info on what you're using. Anyway, I'll remove the `I`s

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/Commands/AddResource.java:
```diff
@@ -0,0 +1,27 @@
+    private String value;
+    private SMSNetworkManager networkManager;
+
+    public AddResource(String key, String value, SMSNetworkManager networkManager){
```
Why are these commands specific to sms? Couldn't you ask for a `NetworkDictionary` and a `NetworkBroadcaster`?

Cogno-Marco replied:
I actually proposed this exact change twice, and I still want this exact change to happen, so yes, I'll do it.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/SMSNetDictionary.java:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @author Marco Cognolato
+ */
+public class SMSNetDictionary implements INetworkDictionary<String, String> {
```
Are you restricting to `String` only for now? Did you take account that one day we could be able to handle whatever type of `Serializable` object? Would you implement the serialization and de-serialization at dictionary level or at manager level?
Not saying that you should to it now, not at all, just asking if you already thought about this and if you needed any kind of help.

Cogno-Marco replied:
String is only for now, yes.
About serialization I'll probably either do it at manager level, or use the Strategy Pattern to let the user do it, I don't have an exact plan as of today. So yes, any help is greatly appreciated 👍

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/SMSNetDictionary.java:
```diff
@@ -0,0 +1,55 @@
+        return dict.get(key);
+    }
+
+    private void checkKeyValidity(String key){
```
Missing specifications, like what's a "valid" key and why is it valid only if it's composed of a single word??

Cogno-Marco replied:
I'll add specifications 👍 
The key is valid only if it's a single word because I'd like the resource to be more than a word, and that is fine at a dictionary level, but when you send the smsmessage key and resource get tangled up, so you can't distinguish where the key ends and where the resource starts (unless you use a specific indicator obv.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/SMSNetSubscribers.java:
```diff
@@ -0,0 +1,43 @@
+ *
+ * @author Marco Cognolato
+ */
+public class SMSNetSubscribers implements INetSubscribers<SMSPeer>  {
```
The "I" in `INetSubscribers` makes me shiver.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/SMSNetSubscribers.java:
```diff
@@ -0,0 +1,43 @@
+     * @param subscriber The subscriber to add to the net
+     * @throws IllegalArgumentException If subscriber is null
+     */
+    public void addSubscriber(SMSPeer subscriber){
```
What about a `@NonNull final` then? To help to avoid null subscribers

Cogno-Marco replied:
`@NonNull` doesn't force any constraint for null values, so I decided to handle it myself. I'll still add the `@NonNull` to let the user know at compile time though 👍

Cogno-IDU commented:
> How about placing it inside the .gitignore?

I think there's a better solution, I remember something about forcing android studio into choosing a specific java version

CremaLuca reviewed (changes requested):
I'm not finished, this has been a quick review, I probably have some more later.

Cogno-Marco committed:
Code review edits (by Luca Crema)
+ forced java version to 1.8
~ Corrected typo in INetSubscribers, SMSNetSubscriberTest and SMSNetSubscribers
~ Commands now require the list they need instead of the whole network manager
~ Fixed tests due to Commands updates
+ Added missing specifications in SMSNetDictionary

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/Commands/AddResource.java:
```diff
@@ -0,0 +1,27 @@
+    private String value;
+    private SMSNetDictionary netDictionary;
+
+    public AddResource(String key, String value, SMSNetDictionary netDictionary){
```
Why `SMS`NetDictionary and not a generic `NetworkDictionary`? Why is this restricted to SMS and if this is wanted, why is it in the `network` package (where there should be generic things)?

CremaLuca replied:
Also missing specifications lol

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/Commands/AddPeer.java:
```diff
@@ -0,0 +1,26 @@
+    private SMSPeer peer;
+    private SMSNetSubscribers netSubscribers;
+
+    public AddPeer(SMSPeer peer, SMSNetSubscribers netSubscribers){
```
As in `AddResource`, why the SMS?

CremaLuca replied:
Also SMSPeer could become a generification by doing `AddPeer<P extends Peer>`

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/Commands/RemoveResource.java:
```diff
@@ -0,0 +1,24 @@
+    private SMSNetDictionary netDictionary;
+    private String key;
+
+    public RemoveResource(String key, SMSNetDictionary netDictionary){
```
As in `AddResource`

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/INetSubscribers.java:
```diff
@@ -0,0 +1,31 @@
+ *
+ * @author Marco Cognolato
+ */
+public interface INetSubscribers<T extends Peer> {
```
I think adding the -s at the end of a class/interface name to use it as a list is not that standard, you usually use singular names, I'd have used `NetSubscriberList/Collection/Store`

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/SMSNetDictionary.java:
```diff
@@ -0,0 +1,60 @@
+package com.eis.communication.network;
```
I'd move every implementation with SMS in a different package. We worked hard to be as generic as possible with the interfaces in `communication.network`, let's not waste what we've done.

Cogno-Marco replied:
My problem is I really struggle to work with generalization, I tend to work on a concrete example and THEN generalize.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/SMSNetDictionary.java:
```diff
@@ -0,0 +1,60 @@
+
+    /**
+     * Checks if a given key is valid, else throws IllegalArgumentException.
+     * A key is said to be valid only if it's composed of one word
```
Why is it so? Say it somewhere so that one is aware of the reason you decided to put this limit.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/INetSubscribers.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.ArrayList;
+
+/**
+ * Interface which defines common operations for Subscribers of a network
```
What exactly is a "Subscriber", what does it represent? It is not defined anywhere.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/Commands/AddResource.java:
```diff
@@ -0,0 +1,27 @@
+import com.eis.communication.network.SMSNetDictionary;
+
+/**
+ * Command to add a resource
```
Add a resource where? What? Why? Who is allowed to add (idk, stupid queston)? Be as specific as you can.

CremaLuca reviewed (changes requested):
Still a lot to change. Also I'd like to see how commands are used, could you maybe implement one method of the `SMSNetworkManager` like addUser or addResource? Because I'm a little concerned about a thing but I don't want to anticipate anything before I see some actual implementation.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/Commands/Command.java:
```diff
@@ -0,0 +1,17 @@
+package com.eis.communication.network.Commands;
+
+/**
+ * This interface has to be implemented by any class that needs to be a command
```
Redundant and useless comment.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/Commands/RemovePeer.java:
```diff
@@ -0,0 +1,26 @@
+    private SMSPeer peer;
+    private SMSNetSubscribers netSubscribers;
+
+    public RemovePeer(SMSPeer peer, SMSNetSubscribers netSubscribers){
```
Same for AddPeer, could have used a `<P extends Peer>`

Cogno-Marco committed:
- Removed I in Interface names
~ Renamed NetSubscribers to NetSubscriberList
+ Moved SMS implementations into proper package
~ renamed packages to smallcase

Cogno-IDU commented:
I'll give the most basic example of a command implementation.
AddPeer has to 
1. add a peer to the local subscribers list
2. broadcast that the peer was added

implementation of this will be as simple as
```
netSubscribers.addSubscriber(peer);
String message = RequestType.AddPeer.ordinal() + " " + peer.toString();
BroadcastModule.broadcast(message);
```
N.B. I'm anticipating things here, names and some stuff might change, but it still should be along these lines

Cogno-Marco committed:
Moved SMS implementations into proper folders
~ changed folder structure to have 2 packages: one generic and one specific for an sms network
~ Fixed import errors
~ Improved some documentation for Commands
~ Improved Documentation for SMSNetDictionary key validity

Cogno-Marco committed:
~ Improved documentation for the 4 basic commands
+ Added Marco Cognolato as a commands author

Cogno-IDU commented:
Ok, this should be everything (for now)

Cogno-Marco committed:
Improved documentation for common Command interface

Cogno-Marco committed:
Fixed import errors

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/CommandExecutor.java:
```diff
@@ -0,0 +1,18 @@
+ */
+public class CommandExecutor <C extends Command>{
+
+    /**
```
```suggestion
    /**
    *   Calls {@link Command#execute} parameter's method.
```

Cogno-Marco committed:
Removed overgeneralyzation

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSJoinableNetManager.java:
```diff
@@ -0,0 +1,38 @@
+
+    /**
+     * Accepts a given join invitation.
+     * If a listener is NOT set (using {@link #setJoinInvitationListener(JoinInvitationListener)})
```
I'd add this explanation in the class description as one might not read every method specification when using the class.

CremaLuca replied:
You can copy it and leave both explaination.
Two is megl che uan

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetSubscriberList.java:
```diff
@@ -0,0 +1,45 @@
+     * @param subscriber The subscriber to add to the net
+     * @throws IllegalArgumentException If subscriber is null
+     */
+    public void addSubscriber(SMSPeer subscriber){
```
```suggestion
    public void addSubscriber(@NonNull final SMSPeer subscriber){
```

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetSubscriberList.java:
```diff
@@ -0,0 +1,45 @@
+     * @param subscriber The subscriber to remove
+     * @throws IllegalArgumentException If a non present subscriber is removed
+     */
+    public void removeSubscriber(SMSPeer subscriber){
```
```suggestion
    public void removeSubscriber(@NonNull final SMSPeer subscriber){
```

Cogno-Marco replied:
I get the NonNull, but why the final?

CremaLuca replied:
It's a stricter contract with the caller of the method to say that the method is not going to edit the parameter.
https://stackoverflow.com/a/18156816
>Some programmers (like me) think that's a very good thing and use final on almost every parameter. It makes it easier to understand a long or complex method (though one could argue that long and complex methods should be refactored.) It also shines a spotlight on method parameters that aren't marked with `final`.

Cogno-Marco replied:
oh alright, good to know!

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -0,0 +1,90 @@
+/**
+ * The manager class of the network.
+ *
+ * @author Edoardo Raimondi, Marco Cognolato
```
```suggestion
 * @author Edoardo Raimondi
 * @author Marco Cognolato
```

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * @return netSubscribers
+     */
+    public SMSNetSubscriberList getNetSubscribers(){ return netSubscribers; }
```
```suggestion
    public SMSNetSubscriberList getNetSubscriberList(){ return netSubscribers; }
```

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/commands/AddPeer.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Command to add a peer to the Subscribers list
+ *
+ * @author Edoardo Raimondi, Marco Cognolato
```
```suggestion
 * @author Edoardo Raimondi
 * @author Marco Cognolato
```

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/commands/AddResource.java:
```diff
@@ -0,0 +1,39 @@
+package com.eis.smsnetwork.commands;
```
Why are these commands restricted to SMS? Couldn't we have
```Java
public class AddResource<D extends NetDictionary, B extends NetBroadcaster> implements Command{}

public class com.eis.smsnetwork.commands.AddResource extends com.eis.communication.network.AddResource<SMSNetDictionary, SMSNetBroadcaster>{}
```
so that every implementation can use the same command class.

Also, there is no need for the second class (that would become empty, just a placeholder), inside ~~`NetworkDictionary`~~ you could use `CommandExecutor(new AddResource<SMSNetDictionary, SMSNetBroadcaster>(dict, broadcaster))`

Edit: I meant inside `NetworkManager` you could use ...

Cogno-Marco replied:
wait, why in NetworkDictionary?

CremaLuca replied:
I don't know, I thought that's where these commands will be created, am I wrong?

Cogno-Marco replied:
Also yeah, we could totally have generics instead of restricting to SMS messages, but I'd also like to have something working before doing so.

CremaLuca replied:
Using generic would make it still "working". Also now that I have had some more time to think, there is no need for generics, you could pass an interface of `NetDictionary` as a parameter and call `dict.addResource()` it would still work

CremaLuca replied:
```Java
public class AddResource implements Command {

    private String key;
    private String value;
    private NetDictionary netDictionary;

    /**
     * Constructor for the AddResource command, needs the data to operate
     * @param key The key of the resource to add
     * @param value The value of the resource to add
     * @param netDictionary The dictionary to add the resource in
     */
    public AddResource(@NonNull String key ,@NonNull String value, @NonNull NetDictionary netDictionary){
        this.key = key;
        this.value = value;
        this.netDictionary = netDictionary;
    }

    /**
     * Adds the key-resource pair to the dictionary, then broadcasts the message
     */
    public void execute() {
        netDictionary.addResource(key, value);
        //TODO broadcast
    }
}
```

This does the same exact job and is not restricted to SMS

Cogno-Marco replied:
the network dictionary just keeps the resources, it's just a data structure, it shouldn't do any work outside of that.
Commands get created where requests starts, as in methods of NetworkManager
(`invite`, `addPeer`, ...)

CremaLuca replied:
Oh fuck me, yes, I confused Dictionary and Manager. I meant `NetworkManager` sorry.

Cogno-Marco replied:
Anyway, as I said, we can totally do that! But I'd like to have something working before doing so.
The deadline is getting closer afterall

CremaLuca replied:
Are you aware that applying the changes I proposed is going to let the class exactly as it is right now, do you? It's going to work the same way but with widened pre-conditions.

Cogno-Marco replied:
Oh yeah you're right.
Oke, I'll update it. 👍

CremaLuca replied:
It's just a matter of removing SMS- prefix and moving it inside another package, everything else stays the same (except a renaming action)

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/commands/RemoveResource.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Command to remove a resource from the network dictionary
+ *
+ * @author Edoardo Raimondi, Marco Cognolato
```
JavaDoc prefers multiple @author tags

CremaLuca reviewed (changes requested):
I added some changes to a few files, consider changing the same things in other files too.
Consider also `ctrl+alt+l` every file or check the boxes on the right when committing via Android Studio that do the same thing for you.

Overall we're not there yet but we're getting closer

Cogno-Marco committed:
Improved author tags

Co-Authored-By: Luca Crema <lc.crema@hotmail.com>

Cogno-Marco committed:
Changed function name to reflect class name update

Co-Authored-By: Luca Crema <lc.crema@hotmail.com>

Cogno-Marco committed:
Updated author tags

Co-Authored-By: Luca Crema <lc.crema@hotmail.com>

Cogno-Marco committed:
~ Improved author tags for Commands
~ formatted code for Commands

Cogno-Marco committed:
+ Added nonnull final in SMSNetSubscriberList
~ formatted code on some classes
~ Improved documentation comments in SMSJoinableNetManager
~ Improved documentation comments in CommandExecutor

Cogno-Marco committed:
~ Generalized commands
~ Moved commands to different package
~ Renamed NetworkDictionary to NetDictionary

Cogno-Marco committed:
~ Fixed package creation errors
~ Fixed function renaming errors

CremaLuca reviewed (approved):
Now it looks like something I'd approve. I'm dying to see how the commands will be used in the `NetworkManager` and in the `Broadcaster`. Good job by asking for reasons for the changes I were suggesting.

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -0,0 +1,96 @@
+ */
+public class SMSNetworkManager implements NetworkManager<String, String, SMSPeer, FailReason> {
+
+    private SMSNetSubscriberList netSubscribers = new SMSNetSubscriberList();
```
Does this have to be `SMSNetSubscriberList `, couldn't it be a `NetSubscriberList<SMSPeer>`? So that it's independent of the type of implementation/class used.

Cogno-Marco replied:
well, `SMSNetSubscriberList implements NetSubscriberList<SMSPeer>`, wouldn't that be the same?
Afterall **SMS** net manager would net **SMS** net sub list, right?

CremaLuca replied:
What if I wanted to use my own `NetSubscriberList<SMSPeer>` that orders peers by telephone number and pass it as a parameter of the constructor of `SMSNetworkManager` because I want to keep everything else the same?

Cogno-Marco replied:
my god, you want too much.
what if I add my list which just purposely doesn't store anything?

I get that you want this to be really exapandable etc, but... isn't this too much?

CremaLuca replied:
It is a lot of generalization indeed but doesn't cost too much and makes use of interfaces as real usable things instead of placeholders for methods to be implemented.

Cogno-Marco replied:
Alright, but, in case you forgot, this is gonna be a singleton, so... `SMSNetworkManager.getInstance(yourList)`? `SMSNetworkManager.getInstance().setList(yourList)`?

Cogno-Marco replied:
`setup(yourList, yourDictionary)`?

CremaLuca replied:
Definitely `setNetSubscriberList()` and `setNetDictionary()`

CremaLuca replied:
Is this really going to be a singleton? I couldn't remember the pros and cons

Cogno-Marco replied:
It's going to be a singleton only early on, to have something working really soon. We'll switch to something else (object pool?) when we know it's definitely working

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -0,0 +1,96 @@
+public class SMSNetworkManager implements NetworkManager<String, String, SMSPeer, FailReason> {
+
+    private SMSNetSubscriberList netSubscribers = new SMSNetSubscriberList();
+    private SMSNetDictionary netDictionary = new SMSNetDictionary();
```
Could be `NetDictionary<String, String>`.

Cogno-Marco replied:
Same as the NetSubList, SMSNetDictionary implements NetDictionary<String,String>

Cogno-Marco committed:
~ Formatted class
~ Generalized sub list and dictionary
+ Added methods to provide your implementations of a dictionary and a list

tonin-alessandra commented on networklibrary/src/main/java/com/eis/communication/network/CommandExecutor.java:
```diff
@@ -0,0 +1,20 @@
+package com.eis.communication.network;
+
+/**
+ * Class that manage all the sent commands.
```
Just a little typo, you wrote manage instead of manages

tonin-alessandra commented on networklibrary/src/main/java/com/eis/communication/network/NetDictionary.java:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * Removes a resource from the dictionary
+     * @param key The key which defines the resource
+     */
```
I think you should specify "to be removed" on @param specification, as you wrote in the getResource method.
Same also in addResource method.

tonin-alessandra commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetDictionary.java:
```diff
@@ -0,0 +1,78 @@
+    /**
+     * Adds a resource to the network dictionary
+     *
+     * @param key      The key which defines the resource
```
As said before, maybe it would be better to specify what resource is defined by the key.
Same in the next method.

tonin-alessandra commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetDictionary.java:
```diff
@@ -0,0 +1,78 @@
+     * We decided to use a space as a separator, so if we allow multiple words keys we run into a
+     * problem: understanding how to distinguish a key from a resource.
+     * For example if a message has "a b c" is it the key "a" and the resource "b c", or the
+     * key "a b" and the resource "c".
```
You forgot a question mark at the end of this statement, I think.

tonin-alessandra commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetSubscriberList.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    public void addSubscriber(@NonNull final SMSPeer subscriber) {
+        if (subscriber == null) throw new IllegalArgumentException("Cannot add a null peer!");
+        subscribers.add(subscriber);
```
I have a doubt: since you already wrote @NonNull on the param, is it necessary to check "if (subscriber == null)"?

CremaLuca replied:
@NonNull doesn't prevent null objects to be passed as a parameter, so yes, it's still required.

tonin-alessandra reviewed (approved):
Good idea using the Command pattern, it seems to be really useful; I didn't know about this, I learned something new, thanks!
I think this is a nice code, nothing particular to say, the comments I made are all related to little things on specifications: they are not important changes, so I approve this pull request.
Good job!

Cogno-IDU commented:
Fixed, last typos, I think this can be merged now 👍

Cogno-Marco committed:
~ Fixed typo in CommandExecutor
~ Formatted code in NetDictionary
~ Improved documentation in NetDictionary
~ Fixed typos in SMSNetDictionary

Cogno-Marco committed:
Fixed more typos in SMSNedDictionary
